Manual alarm with MQTT control

colinodell opened the pull request:
This PR allows a manual alarm to be controlled via MQTT.

## Description:

The MQTT alarm component allows HA to control a standard alarm panel via MQTT; commands to arm/disarm the alarm come from HA, and the physical panel determines when to trigger the alarm.

However, I'd like the opposite functionality - I want to create [a "dumb" keypad using a Raspberry Pi](https://twitter.com/colinodell/status/881258798631067648) which arms/disarms the `manual` alarm in HA (via MQTT).  HA would determine if/when to trigger the alarm based on my automations.  The Raspberry Pi would also subscribe to messages about the current state of the alarm and update its display accordingly.

## Implementation

My proposal simply extends the existing `ManualAlarm` class to do 3 things:

 1. Adds the `mqtt` alarm configuration options as well
 2. Subscribes to an MQTT command topic; calls the existing `async_alarm_disarm` / `async_alarm_arm_*` methods when it receives those commands
 3. Tracks its own state changes and publishes MQTT messages with the new state

## Why this approach?

I looked into using existing MQTT functionality and the various APIs, but none of them fit my needs:

 - The MQTT platform does allow me to publish state change messages to the Pi (via custom automations), but it does not seem to provide communication in the opposite direction - the ability to subscribe to MQTT topics and call services as a result.
 - The REST API does not offer the ability to subscribe to state changes - I'd have to DDoS HA from my Pi to get real-time info.
 - The Web Sockets API does offer subscription functionality, but not for single entities.  A Raspberry Pi Zero has relatively low processing power, especially when driving a display, so I'd like to avoid DDoSing the Pi by having to process every single state change from HA.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2921

## Example entry for `configuration.yaml` (if applicable):

```yaml
alarm_control_panel:
  platform: manual_mqtt
  state_topic: "home/alarm"
  command_topic: "home/alarm/set"
  name: Alarm
```

All other configuration options from both `manual` and `mqtt` alarms are also supported.

## Feedback Requested:

I'd greatly appreciate the following feedback:

 - Is this something you'd consider merging into HA?
 - Am I doing the async/callback stuff correctly?
 - Can you suggest a better name than `manual_mqtt` or is that fine?

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

colinodell committed:
Manual alarm with MQTT control

colinodell commented:
I don't know why the `Python: 3.6-dev` test is failing, but I'm pretty sure its not because of my changes.  I'm therefore going to mark tox as passing.

**This is now ready for review**

colinodell commented:
If you're interested in seeing how I'm using this:

 - https://www.youtube.com/watch?v=2Lei8n_aSJI
 - https://github.com/colinodell/mqtt-control-panel

MartinHjelmare commented:
I think it should work to do what you want using the existing mqtt platform. Didn't it work to configure a state topic which the RPi publishes to?

colinodell commented:
> Didn't it work to configure a state topic which the RPi publishes to?

I don't want to set the state directly. Manual alarms provide a configurable delay before the alarm is fully armed (via the `pending_time` configuration parameter).  AFAICT, you need to call the `alarm_control_panel.alarm_arm_*` service for this delay to work, and the existing MQTT platform does not seem to provide that ability.

Or perhaps I'm not fully understanding what you mean - if that's the case, would you mind providing an example of configuring HA to call this service when it receives an MQTT message from the RPi?

colinodell commented:
I've rebased this onto the latest `dev` branch.

MartinHjelmare commented:
Ah, ok, I see what you want to do now. No, that's not possible currently with the existing mqtt platform. I guess the question is, do we want a new platform or an extension of the current mqtt platform?

colinodell commented:
> I guess the question is, do we want a new platform or an extension of the current mqtt platform?

The manual alarm is a special case IMO, so I don't think it's too far-fetched to create a new platform for it.  I tried to make the implementation as simple as possible to reduce ongoing maintenance:

 - The important bits are only 30 lines long (one `if` statement and some `async_` function calls)
 - The existing manual alarm logic is reused (via inheritance)
 - MQTT functionality uses existing methods

If having a new platform is still a concern, I'd be open to adding this extra MQTT functionality directly into the existing `manual` alarm, but only activating it when a certain configuration is set.

I'm also open to extending the MQTT platform, but I'd need a little more info on how you envision that working.  Are you thinking of expanding the general abilities of MQTT so you can do things like call services and subscribe to certain entity state changes?  It's certainly a viable approach, but are there enough other use cases for it?  (Maybe, I don't know)

MartinHjelmare commented:
Thinking some more about this: Since you basically want to couple the manual alarm feature with the mqtt alarm feature, why not use both existing platforms and use automations as the glue. No new platform would be needed and it would be very customizable.

The manual alarm platform would hold the actual state of the alarm while the mqtt alarm platform would just represent the state of the physical "dumb" interface panel. When a user changes state of the panel, mqtt publishes the new state to the mqtt alarm platform, which also changes state. An automation is then triggered that calls a service of the manual alarm platform which in turn changes the state of that platform.

What do you think?

balloob commented:
Sorry have been busy lately with many things so didn't look at this PR thoroughly. I do agree with @MartinHjelmare that we can't just use the manual alarm and make it MQTT proof. That simply doesn't scale.

So instead, just copy paste the manual alarm code and tests and adjust the platform and tests.

colinodell committed:
Duplicate manual control panel code instead of extending it

colinodell committed:
Duplicate manual alarm test as well; modify for manual_mqtt

colinodell committed:
Add MQTT-specific tests for manual_mqtt alarm

colinodell commented on homeassistant/components/alarm_control_panel/manual_mqtt.py:
```diff
@@ -0,0 +1,235 @@
+
+import homeassistant.components.alarm_control_panel as alarm
+import homeassistant.util.dt as dt_util
+from homeassistant.const import (
```
Resolved and force-pushed. Not sure why @houndci-bot hasn't noticed its removal.

balloob replied:
It's not that smart.

colinodell commented:
Thanks for the feedback @balloob!

As request, I've copypasted the `manual` alarm and added my changes.  I did the same for the tests and also added tests for the MQTT functionality.

balloob commented:
🐬  thanks !
